app/uk/gov/hmrc/agentmappingfrontend/controllers/MappingController.scala

```diff
@@ -58,32 +62,55 @@ class MappingController @Inject()(
 
   def startSubmit: Action[AnyContent] = Action.async { implicit request =>
     withAuthorisedAgent {
-      successful(Redirect(routes.MappingController.showAddCode()))
+      successful(Redirect(routes.MappingController.showEnterAccountNo))
     }
   }
 
-  def showAddCode: Action[AnyContent] = Action.async { implicit request =>
+  def showEnterAccountNo: Action[AnyContent] = Action.async { implicit request =>
     withAuthorisedAgentAudited {
-      successful(Ok(html.add_code(mappingForm)))
+      successful(Ok(html.enter_account_number(mappingFormArn)))
     }(AuditService.auditCheckAgentRefCodeEvent(auditService))
   }
 
-  def submitAddCode: Action[AnyContent] = Action.async { implicit request =>
+  def submitEnterAccountNo: Action[AnyContent] = Action.async { implicit request =>
+    withAuthorisedAgent {
+      mappingFormArn.bindFromRequest.fold(
+        formWithErrors => {
+          successful(Ok(html.enter_account_number(formWithErrors)))
+        },
+        formArn => {
+          successful(Redirect(routes.MappingController.showEnterUtr).addingToSession(("mappingArn", formArn.arn.value)))
+        }
+      )
+    }
+  }
+
+  def showEnterUtr: Action[AnyContent] = Action.async { implicit request =>
+    withAuthorisedAgent {
+      request.session.get("mappingArn").isDefined match {
+        case true  => successful(Ok(html.enter_utr(mappingFormUtr)))
+        case false => successful(Redirect(routes.MappingController.showEnterAccountNo()))
+      }
+    }
+  }
+
+  def submitEnterUtr: Action[AnyContent] = Action.async { implicit request =>
     withAuthorisedAgent {
-      mappingForm.bindFromRequest.fold(
+      mappingFormUtr.bindFromRequest.fold(
         formWithErrors => {
-          successful(Ok(html.add_code(formWithErrors)))
+          successful(Ok(html.enter_utr(formWithErrors)))
         },
-        mappingData => {
-          mappingConnector.createMapping(mappingData.utr, mappingData.arn) map { r: Int =>
-            r match {
-              case CREATED => Redirect(routes.MappingController.complete())
-              case FORBIDDEN =>
-                Ok(
-                  html.add_code(mappingForm.withGlobalError(
-                    "These details don't match our records. Check your account number and tax reference.")))
-              case CONFLICT => Redirect(routes.MappingController.alreadyMapped())
-            }
+        formWithUtr => {
+          request.session.get("mappingArn") match {
+            case Some(arn) =>
+              mappingConnector.createMapping(Utr(formWithUtr.utr.value), Arn(arn)) map { r: Int =>
```

Comment:
you can remove `r:Int` from here. Please rename `r` to status or response.

Reply:
The type of `formWithUtr.utr` is Utr, right? In that case you can avoid enclosing that in Utr here.